<a href="https://colab.research.google.com/github/FeliksMakarios/ml-model_project/blob/main/UL_Analisis_Rating_Mie_Instan.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Analisis Rating Mie Instan**

Project ini akan mengembangkan model machine learning untuk memprediksi rasa mie terbaik yang akan dijadikan ide selanjutnya untuk dijual. Untuk melakukannya, perlu melakukan analisis pada dataset rating mie instan.

Dataset yang digunakan berasal dari [Ramen Rater](https://www.theramenrater.com/resources-2/the-list/). Ramen Rater merupakan blog yang memberikan ulasan pada semua mie di seluruh dunia sejak tahun 2002. Mereka memberikan dataset yang terbuka untuk publik.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# import library os untuk bisa berinteraksi dengan file system yang ada pada operating system (os)
import os

os.chdir("/content/drive/My Drive/Dataset/Ramen Rating Analysis")

## **Pengumpulan dan Pembersihan Data**

### Pengumpulan Data

In [3]:
import pandas as pd

In [17]:
# baca file excel sebagai dataframe
df = pd.read_excel('The-Big-List-20210117.xlsx')
df

Review #            Brand  ... Stars   T
0         3702      Higashimaru  ...     5 NaN
1         3701     Single Grain  ...   3.5 NaN
2         3700          Sau Tao  ...     5 NaN
3         3699          Sau Tao  ...   4.5 NaN
4         3698  Sapporo Ichiban  ...   3.5 NaN
...        ...              ...  ...   ...  ..
3697         5            Vifon  ...   3.5 NaN
3698         4          Wai Wai  ...     1 NaN
3699         3          Wai Wai  ...     2 NaN
3700         2          Wai Wai  ...     2 NaN
3701         1         Westbrae  ...   0.5 NaN

[3702 rows x 7 columns]

### Pembersihan Data

In [18]:
# memeriksa kolom mana saja yang mengandung missing values
df.isnull().sum()

Review #       0
Brand          0
Variety        0
Style          0
Country        0
Stars          0
T           3702
dtype: int64

In [19]:
# buang kolom yang tidak dibutuhkan. dalam kasus ini yang akan dibuang adalah kolom Review dan T
df = df.drop(['Review #', 'T'], axis=1)
df

Brand  ... Stars
0         Higashimaru  ...     5
1        Single Grain  ...   3.5
2             Sau Tao  ...     5
3             Sau Tao  ...   4.5
4     Sapporo Ichiban  ...   3.5
...               ...  ...   ...
3697            Vifon  ...   3.5
3698          Wai Wai  ...     1
3699          Wai Wai  ...     2
3700          Wai Wai  ...     2
3701         Westbrae  ...   0.5

[3702 rows x 5 columns]

In [20]:
df.dtypes

Brand      object
Variety    object
Style      object
Country    object
Stars      object
dtype: object

Memeriksa lebih jauh isi dari kolom Stars. mengapa tipenya object

In [21]:
# cek nilai unik yang ada pada kolom Stars dan tampilkan dalam bentuk arrays
df['Stars'].unique()

array([5, 3.5, 4.5, 4, 3.75, 4.25, 3, 3.25, 4.75, 2.5, 2, 0.75, 0, 1.25,
       2.75, 0.5, 1.5, 2.25, 1, 'NS', 0.25, 'NR', 1.75, '3.5/2.5',
       datetime.datetime(2017, 4, 4, 0, 0),
       datetime.datetime(2017, 5, 5, 0, 0), '4.5/5', '5/2.5',
       datetime.datetime(2017, 5, 4, 0, 0), '4.25/5', 'Unrated', 1.1, 2.1,
       0.9, 3.1, 4.125, 3.125, 2.125, 2.9, 0.1, 2.8, 3.7, 3.4, 3.6, 2.85,
       2.3, 3.2, 3.65, 1.8], dtype=object)

In [22]:
# cek nilai unik yang ada pada kolom Stars dan tampilkan dalam bentuk series
df['Stars'].value_counts()

5                      667
3.5                    543
3.75                   535
4                      439
4.5                    242
3.25                   222
4.25                   215
3                      172
2.75                   110
2                       95
4.75                    93
2.5                     93
1.5                     50
0                       40
1                       38
2.25                    29
1.75                    28
0.5                     20
0.25                    17
1.25                    14
Unrated                  3
NR                       3
4.25/5                   2
0.75                     2
4.125                    2
3.1                      2
2.9                      2
2.8                      2
1.1                      2
2.1                      1
5/2.5                    1
3.4                      1
2017-04-04 00:00:00      1
2.3                      1
3.2                      1
1.8                      1
0.1                      1
3

Hasilnya menunjukkan isi dari kolom Stars sangat bervariasi. Seharusnya hanya ada data dari 1-5 tapi ternyata ada isi lainnya yang kurang sesuai seperti `4.5/5` atau dalam bentuk datetimes `2017-05-05 00:00:00` atau juga dalam bentuk string `NS`. Nilai-nilai ini perlu diganti/replace.

In [23]:
# 4.5/5 -> 4.5
# 5/2.5 -> 2.5
# 3.5/2.5 -> 2.5
# 4.25/5 -> 4.25

df['Stars'] = df['Stars'].replace('4.5/5', 4.5)
df['Stars'] = df['Stars'].replace('5/2.5', 2.5)
df['Stars'] = df['Stars'].replace('3.5/2.5', 2.5)
df['Stars'] = df['Stars'].replace('4.25/5', 4.5)

df['Stars'].value_counts()

5                      667
3.5                    543
3.75                   535
4                      439
4.5                    245
3.25                   222
4.25                   215
3                      172
2.75                   110
2                       95
2.5                     95
4.75                    93
1.5                     50
0                       40
1                       38
2.25                    29
1.75                    28
0.5                     20
0.25                    17
1.25                    14
NR                       3
Unrated                  3
0.75                     2
2.8                      2
4.125                    2
1.1                      2
2.9                      2
3.1                      2
2.85                     1
2017-04-04 00:00:00      1
2.3                      1
3.2                      1
1.8                      1
0.1                      1
3.6                      1
3.4                      1
3.7                      1
3

Beberapa data sudah terganti. Selanjutnya nilai yang dalam bentuk string akan dibuang. Namun, terlebih dahulu nilai-nilai tersebut harus dibuat menjadi `NaN` agar nanti lebih mudah dibuang. Untuk membuat nilai selain integer/float menjadi `NaN`, kita dapat melakukan konversi tipe data untuk kolom `Stars`.

In [24]:
# jika methode to_numeric menemukan string atau apapun yg bukan number, maka dia akan mengubahnya menjadi NaN
df['Stars'] = pd.to_numeric(df['Stars'], errors= 'coerce')
df['Stars'].value_counts()

5.000    667
3.500    543
3.750    535
4.000    439
4.500    245
3.250    222
4.250    215
3.000    172
2.750    110
2.000     95
2.500     95
4.750     93
1.500     50
0.000     40
1.000     38
2.250     29
1.750     28
0.500     20
0.250     17
1.250     14
1.100      2
3.100      2
2.900      2
0.750      2
2.800      2
4.125      2
2.850      1
2.300      1
3.600      1
3.700      1
3.650      1
0.100      1
3.400      1
3.125      1
1.800      1
3.200      1
2.100      1
2.125      1
0.900      1
Name: Stars, dtype: int64

In [25]:
# periksa lagi apakah semua kolom masih mengandung NaN
df.isnull().sum()

Brand       0
Variety     0
Style       0
Country     0
Stars      10
dtype: int64

Ada 10 data pada kolom `Stars` yang dianggap NaN. Sepuluh ini adalah:
* NR = 3
* Unrated = 3
* 2017-04-04 00:00:00 = 1
* 2017-05-04 00:00:00 = 1
* 2017-05-05 00:00:00 = 1
* NS = 1

Sepuluh nilai NaN ini selanjutnya akan dibuang.

In [26]:
# buang semua baris yang memiliki NaN didalamnya
df = df.dropna()

df.isnull().sum()

Brand      0
Variety    0
Style      0
Country    0
Stars      0
dtype: int64

In [27]:
df

Brand  ... Stars
0         Higashimaru  ...   5.0
1        Single Grain  ...   3.5
2             Sau Tao  ...   5.0
3             Sau Tao  ...   4.5
4     Sapporo Ichiban  ...   3.5
...               ...  ...   ...
3697            Vifon  ...   3.5
3698          Wai Wai  ...   1.0
3699          Wai Wai  ...   2.0
3700          Wai Wai  ...   2.0
3701         Westbrae  ...   0.5

[3692 rows x 5 columns]

Data sudah bersih dan saat ini ada 3692 baris data. Data yang sudah bersih ini selanjutnya perlu di export menjadi CSV agar dapat digunakan lagi suatu waktu.

In [28]:
df.to_csv('thebiglist_clean.csv', index=None)

In [29]:
pd.read_csv('thebiglist_clean.csv')

Brand  ... Stars
0         Higashimaru  ...   5.0
1        Single Grain  ...   3.5
2             Sau Tao  ...   5.0
3             Sau Tao  ...   4.5
4     Sapporo Ichiban  ...   3.5
...               ...  ...   ...
3687            Vifon  ...   3.5
3688          Wai Wai  ...   1.0
3689          Wai Wai  ...   2.0
3690          Wai Wai  ...   2.0
3691         Westbrae  ...   0.5

[3692 rows x 5 columns]

## **Ekstraksi Fitur, Visualisasi Data, dan Analisis Statistik**

### Ekstraksi Fitur

### Visualisasi Data

### Analisis Statistik

## **Rekayasa Fitur dan Persiapan Data**

## **Melatih Model Machine Learning**